integration/experiment/system_characterization/sys_charac_utils.sh

```diff
@@ -0,0 +1,111 @@
+
+launch_sweep () {
+
+          geopmlaunch pals \
+            -n ${RANK_COUNT} -ppn ${RANK_COUNT} \
+            --cpu-bind list:0-${CPU_LOGICAL_CORE_CNT_MINUS_1} \
+            --geopm-init-control=$INIT_CONTROLS_LIST \
+            --geopm-ctl=application \
+            --geopm-preload \
+            --geopm-profile="${PROGRAM_NAME}" \
+            --geopm-report=${REPORT_FILE_PATH} \
+            --geopm-report-signals=${GEOPM_SIGNALS} \
+            --geopm-program-filter=${PROGRAM_NAME} \
+            -- ${BINARY_PATH_PLUS_FLAGS} 2>&1 \
+            > ${LOG_FILE_PATH}
+
+          sleep 5
+}
+
+
+launch_cpu_sweep () {
+
+                RANK_COUNT=${CPU_PHYSICAL_CORE_CNT_MINUS_4}
+                CORE_MIN_FREQ=$(geopmread CPU_FREQUENCY_MIN_AVAIL board 0)
+                CORE_MAX_FREQ=$(geopmread CPU_FREQUENCY_MAX_AVAIL board 0)
+                CORE_FREQ_STEP=$(geopmread CPU_FREQUENCY_STEP board 0)
+                UNCORE_MIN_FREQ=$(geopmread CPU_UNCORE_FREQUENCY_MIN_CONTROL board 0)
+                UNCORE_MAX_FREQ=$(geopmread CPU_UNCORE_FREQUENCY_MAX_CONTROL board 0)
+                UNCORE_FREQ_STEP=100000000
+                GEOPM_SIGNALS="${EXTRA_SIGNALS},MSR::QM_CTR_SCALED_RATE@package,CPU_UNCORE_FREQUENCY_STATUS@package,MSR::CPU_SCALABILITY_RATIO@package,CPU_FREQUENCY_MIN_CONTROL@package,CPU_UNCORE_FREQUENCY_MIN_CONTROL@package"
+        
+                for ((p="$CORE_MIN_FREQ"; p<="$CORE_MAX_FREQ"; p=p+"$CORE_FREQ_STEP")); do
+                    for ((u="$UNCORE_MIN_FREQ"; u<="$UNCORE_MAX_FREQ"; u=u+"$UNCORE_FREQ_STEP")); do
+                
+                      printf "${EXTRA_CONTROLS}" > $INIT_CONTROLS_LIST
+                      printf "%s\n" \
+                             "MSR::PQR_ASSOC:RMID board 0 0" \
+                             "MSR::QM_EVTSEL:RMID board 0 0" \
+                             "MSR::QM_EVTSEL:EVENT_ID board 0 2" \
+                             "CPU_FREQUENCY_MIN_CONTROL board 0 ${p}" \
+                             "CPU_FREQUENCY_MAX_CONTROL board 0 ${p}" \
+                             "CPU_UNCORE_FREQUENCY_MIN_CONTROL board 0 ${u}" \
+                             "CPU_UNCORE_FREQUENCY_MAX_CONTROL board 0 ${u}" \
+                             >> $INIT_CONTROLS_LIST
+        
+                      for ((t=0; t<"$TRIAL_COUNT"; t++)); do
+                
+                          echo "=== CPU POWER-FREQ SWEEP: Trial $t, CPU CORE $p, CPU UNCORE $u, BOARD POWER $l==="
+                          REPORT_FILE_PATH="${EMPTY_SWEEP_OUTPUT_DIR}/${PROGRAM_NAME}_boardcap_${l}_core_${p}_uncore_${u}_trial_${t}_cpufreqsweep-${HOSTNAME}.report"
+                          LOG_FILE_PATH="${EMPTY_SWEEP_OUTPUT_DIR}/${PROGRAM_NAME}_boardcap_${l}_core_${p}_uncore_${u}_trial_${t}_cpufreqsweep-${HOSTNAME}.log"
+
+                          launch_sweep
+                
+                      done
+                    done
+                done
+
+}
+
+
+launch_gpu_sweep () {
+
+                RANK_COUNT=1
```

Comment:
Include some comments about the chosen rank count per node:
1. How you arrived at this configuration.
2. Why it's appropriate for this type of sweep.